src/ui/components/SecondaryToolbox/react-devtools/ReactDevToolsListData.ts

```diff
@@ -0,0 +1,92 @@
+import assert from "assert";
+
+import { GenericListData } from "replay-next/components/windowing/GenericListData";
+import { StoreWithInternals } from "ui/components/SecondaryToolbox/react-devtools/ReplayWall";
+import { ReactElement } from "ui/components/SecondaryToolbox/react-devtools/types";
+
+export class ReactDevToolsListData extends GenericListData<ReactElement> {
+  private collapsedFiberIds: Set<number> = new Set();
+  private store: StoreWithInternals;
+
+  constructor(store: StoreWithInternals) {
+    super();
+
+    this.collapsedFiberIds = new Set();
+    this.store = store;
+  }
+
+  getItemById(id: number): ReactElement | null {
+    return this.store.getElementByID(id) as ReactElement | null;
+  }
+
+  processMutatedStore() {
+    // Restore collapse state then selected element state
+    this.collapsedFiberIds.forEach(id => {
+      if (this.store.containsElement(id)) {
+        this.store.toggleIsCollapsed(id, true);
+      } else {
+        this.collapsedFiberIds.delete(id);
+      }
+    });
+
+    let selectedElement = null;
+
+    const selectedIndex = this.getSelectedIndex();
+    if (selectedIndex != null) {
+      selectedElement = this.getItemAtIndex(selectedIndex);
+    }
+
+    this.invalidate();
+
+    if (selectedElement) {
+      this.selectElement(selectedElement);
+    }
+  }
+
+  selectElement(element: ReactElement | null) {
+    if (element) {
+      const index = this.getIndexForItem(element);
+      if (index >= 0) {
+        this.setSelectedIndex(index);
+        return;
+      }
+    }
+
+    this.setSelectedIndex(null);
+  }
+
+  toggleCollapsed(element: ReactElement) {
+    const isCollapsed = !element.isCollapsed;
+
+    if (isCollapsed) {
+      this.collapsedFiberIds.add(element.id);
+    } else {
+      this.collapsedFiberIds.delete(element.id);
+    }
+
+    this.store.toggleIsCollapsed(element.id, isCollapsed);
+    this.invalidate();
+  }
+
+  updateStore(store: StoreWithInternals) {
```

Comment:
:question: Where/when would we need to swap the store instance?

Reply:
Good question. I wrote this early on, thinking that maybe we would need to recreate the store rather than update it. I just forgot to remove the method. (430e47596)